### ISSUE TITLE
Support for Python coroutines

### ISSUE THREAD

**Issue by ovekaaven** [sphinx-doc/sphinx#4777]
Sphinx does not currently seem to support Python coroutines (async functions or methods).

I've written a patch that fixes this, although I'm not sure if you'd prefer a different design. (If you do think it looks good, I can make a pull request.)

- the pycode/parser.py change fixes 'bysource' sorting of coroutines
- the domains/python.py (and doc/domains.rst) change adds an 'async' option, which can be used with py:function, py:method, etc.
- the rest adds autodoc support

In the output, "async" will be prepended to the signature prefix (i.e. where thinds like "static" and "classmethod" show up) if a function or method is a coroutine.

```
diff --git a/doc/domains.rst b/doc/domains.rst
index 5bed02cf4..975f1f7ad 100644
--- a/doc/domains.rst
+++ b/doc/domains.rst
@@ -299,6 +299,17 @@ you can use brackets to specify the optional parts:
 It is customary to put the opening bracket before the comma.
 
 
+.. _coroutines:
+
+Coroutines
+~~~~~~~~~~
+
+If a function or method is a coroutine, you can add the ``:async:`` option::
+
+   .. py:function:: compile(source : string, filename, symbol='file') -> ast object
+      :async:
+
+
 .. _info-field-lists:
 
 Info field lists
diff --git a/sphinx/domains/python.py b/sphinx/domains/python.py
index b0900b385..e9b65801c 100644
--- a/sphinx/domains/python.py
+++ b/sphinx/domains/python.py
@@ -170,6 +170,7 @@ class PyObject(ObjectDescription):
         'noindex': directives.flag,
         'module': directives.unchanged,
         'annotation': directives.unchanged,
+        'async': directives.flag,
     }
 
     doc_field_types = [
@@ -253,6 +254,8 @@ class PyObject(ObjectDescription):
         signode['fullname'] = fullname
 
         sig_prefix = self.get_signature_prefix(sig)
+        if 'async' in self.options:
+            sig_prefix = 'async ' + sig_prefix
         if sig_prefix:
             signode += addnodes.desc_annotation(sig_prefix, sig_prefix)
 
diff --git a/sphinx/ext/autodoc/__init__.py b/sphinx/ext/autodoc/__init__.py
index 89ca1f44c..3ced7ab70 100644
--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -32,7 +32,7 @@ from sphinx.util import rpartition, force_decode
 from sphinx.util.docstrings import prepare_docstring
 from sphinx.util.inspect import Signature, isdescriptor, safe_getmembers, \
     safe_getattr, object_description, is_builtin_class_method, \
-    isenumattribute, isclassmethod, isstaticmethod, getdoc
+    isenumattribute, isclassmethod, isstaticmethod, isasync, getdoc
 
 if False:
     # For type annotation
@@ -1043,6 +1043,13 @@ class FunctionDocumenter(DocstringSignatureMixin, ModuleLevelDocumenter):  # typ
         args = args.replace('\\', '\\\\')
         return args
 
+    def add_directive_header(self, sig):
+        # type: (unicode) -> None
+        Documenter.add_directive_header(self, sig)
+        if isasync(self.object):
+            sourcename = self.get_sourcename()
+            self.add_line(u'   :async:', sourcename)
+
     def document_members(self, all_members=False):
         # type: (bool) -> None
         pass
@@ -1305,6 +1312,13 @@ class MethodDocumenter(DocstringSignatureMixin, ClassLevelDocumenter):  # type:
         args = args.replace('\\', '\\\\')
         return args
 
+    def add_directive_header(self, sig):
+        # type: (unicode) -> None
+        Documenter.add_directive_header(self, sig)
+        if isasync(self.object):
+            sourcename = self.get_sourcename()
+            self.add_line(u'   :async:', sourcename)
+
     def document_members(self, all_members=False):
         # type: (bool) -> None
         pass
diff --git a/sphinx/pycode/parser.py b/sphinx/pycode/parser.py
index f15da664a..315bb267a 100644
--- a/sphinx/pycode/parser.py
+++ b/sphinx/pycode/parser.py
@@ -361,6 +361,8 @@ class VariableCommentPicker(ast.NodeVisitor):
             self.context.pop()
             self.current_function = None
 
+    visit_AsyncFunctionDef = visit_FunctionDef
+
 
 class DefinitionFinder(TokenProcessor):
     def __init__(self, lines):
diff --git a/sphinx/util/inspect.py b/sphinx/util/inspect.py
index e8427fd84..69c72d557 100644
--- a/sphinx/util/inspect.py
+++ b/sphinx/util/inspect.py
@@ -197,6 +197,19 @@ def isdescriptor(x):
     return False
 
 
+try:
+    import asyncio
+except ImportError:
+    asyncio = None
+
+
+def isasync(x):
+    """Check if the object is asynchronous."""
+    if asyncio is None:
+        return False
+    return asyncio.iscoroutinefunction(x)
+
+
 def safe_getattr(obj, name, *defargs):
     # type: (Any, unicode, unicode) -> object
     """A getattr() that turns all exceptions into AttributeErrors."""

```

**Comment by tk0miya** [comment]
+1; I like it. I'd like to merge this after adding testcases.

@shimizukawa what do you think?

**Comment by shimizukawa** [comment]
+1. supporting async keyword is necessary.

**Comment by spectras** [comment]
+1. That's much needed!

**Comment by njsmith** [comment]
You should also be aware of https://github.com/sphinx-doc/sphinx/issues/3743 and [sphinxcontrib-trio](https://sphinxcontrib-trio.readthedocs.io/en/latest/). If you were in interested in working on integrating sphinxcontrib-trio into Sphinx itself then that would be awesome, but even if not then you'll probably want to at least look at it :-).

BTW, I suggest Sphinx should mark async functions with `await` (how they're used) instead of `async` (how they're defined). The advantage is that the `await` approach is more general: you can also use it for documenting decorators (`@foo(...)`), generators (`for ... in foo(...)`), context managers (`with foo(...) as ...`, and even combinations (e.g. an async function returning a sync context manager: `with await foo(...) as ...`).

**Comment by tk0miya** [comment]
I think adding `:async:` option is a part of #3743's approach. So it seems they are not conflicted. Of course, we're welcome your PR to integrate the trio extension into sphinx-core.

BTW, about `await` and `async`, I don't have comment because I've never used asyncio features...

**Comment by tk0miya** [comment]
Fixed by #6295 instead.
Thank you for proposal.
